controllers/topology/machineset_controller.go

```diff
@@ -0,0 +1,211 @@
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// Set finalizer, if it's not already set.
+	if !controllerutil.ContainsFinalizer(ms, clusterv1.MachineSetTopologyFinalizer) {
+		patchHelper, err := patch.NewHelper(ms, r.Client)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to create patch helper for %s", KRef{Obj: ms})
+		}
+		controllerutil.AddFinalizer(ms, clusterv1.MachineSetTopologyFinalizer)
+		if err := patchHelper.Patch(ctx, ms); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to patch %s", KRef{Obj: ms})
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Handle deletion reconciliation loop.
+	if !ms.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, ms)
+	}
+
+	// Nothing to do.
+	return ctrl.Result{}, nil
+}
+
+func (r MachineSetReconciler) reconcileDelete(ctx context.Context, ms *clusterv1.MachineSet) (ctrl.Result, error) {
+	// Calculate the name of the corresponding MachineDeployment.
+	mdName, err := calculateMachineDeploymentName(ms)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Get the corresponding MachineSets.
+	msList, err := getMachineSetsForDeployment(ctx, r.Client, *mdName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the MachineDeployment instance, if it still exists.
+	// Note: This can happen because MachineDeployments are deleted before their corresponding MachineSets.
+	md := &clusterv1.MachineDeployment{}
+	if err := r.Client.Get(ctx, *mdName, md); err != nil {
+		if !apierrors.IsNotFound(err) {
+			// Error reading the object - requeue the request.
+			return ctrl.Result{}, errors.Wrapf(err, "failed to get MachineDeployment/%s", mdName.Name)
+		}
+		// If the MachineDeployment doesn't exist anymore, set md to nil, so we can handle that case correctly below.
+		md = nil
+	}
+
+	// Calculate which templates are still in use by MachineDeployments or MachineSets which are not in deleting.
+	templatesInUse, err := templateutil.CalculateTemplatesInUse(md, msList)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Delete unused templates of the MachineSet.
+	if err := r.deleteUnusedMachineSetTemplates(ctx, templatesInUse, ms); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Remove the finalizer so the MachineSet can be garbage collected by Kubernetes.
+	patchHelper, err := patch.NewHelper(ms, r.Client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to create patch helper for %s", KRef{Obj: ms})
+	}
+	controllerutil.RemoveFinalizer(ms, clusterv1.MachineSetTopologyFinalizer)
+	if err := patchHelper.Patch(ctx, ms); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to patch %s", KRef{Obj: ms})
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// calculateMachineDeploymentName calculates the MachineDeployment name based on owner references.
+func calculateMachineDeploymentName(ms *clusterv1.MachineSet) (*types.NamespacedName, error) {
```

Comment:
I intentionally didn't name it get, because this implies to me some kind of client.Get call we often have in those funcs. But it's fine for me to rename